fx_compile.py

```diff
@@ -0,0 +1,432 @@
+        args_iter = iter(args)
+        env : Dict[str, Node] = {}
+
+        def load_arg(a):
+            return torch.fx.node.map_arg(a, lambda n: env[n.name])
+
+        def fetch_attr(target : str):
+            target_atoms = target.split('.')
+            attr_itr = self.mod
+            for i, atom in enumerate(target_atoms):
+                if not hasattr(attr_itr, atom):
+                    raise RuntimeError(f"Node referenced nonexistant target {'.'.join(target_atoms[:i])}")
+                attr_itr = getattr(attr_itr, atom)
+            return attr_itr
+
+        for node in self.graph.nodes:
+            if node.op == 'placeholder':
+                result = next(args_iter)
+            elif node.op == 'get_attr':
+                result = fetch_attr(node.target)
+            elif node.op == 'call_function':
+                result = node.target(*load_arg(node.args), **load_arg(node.kwargs))
+            elif node.op == 'call_method':
+                self_obj, *args = load_arg(node.args)
+                kwargs = load_arg(node.kwargs)
+                result = getattr(self_obj, node.target)(*args, **kwargs)
+            elif node.op == 'call_module':
+                result = self.modules[node.target](*load_arg(node.args), **load_arg(node.kwargs))
+            elif node.op == 'output':
+                return self.shape_env
+
+            if isinstance(result, torch.Tensor):
+                self.shape_env[node.name] = (result.shape, result.dtype)
+                node.shape = result.shape
+                node.dtype = result.dtype
+
+            env[node.name] = result
+
+        return self.shape_env
+
+
+
+############ New Stuff ##############
+
+def binary_mapping(op):
+    def f(a, b):
+        return op(a, b)
+    return f
+
+decomposition_rules = {}
+binary_decompositions = [
+    (operator.matmul, torch.mm),
+    (operator.add, torch.add),
+    (operator.mul, torch.mul),
+    (operator.sub, torch.sub),
+    (operator.truediv, torch.div),
+    (operator.eq, torch.eq),
+    (operator.gt, torch.gt),
+    (operator.ge, torch.ge),
+    (operator.lt, torch.lt),
+    (operator.le, torch.le),
+    (operator.ne, torch.ne),
+    (operator.and_, torch.bitwise_and)
+]
+for old, new in binary_decompositions:
+    decomposition_rules[old] = binary_mapping(new)
+
+def addmm_decompose(input, mat1, mat2):
+    return torch.add(input , torch.mm(mat1, mat2))
+
+decomposition_rules[torch.addmm] = addmm_decompose
+
+def decompose(model: torch.nn.Module, shape_env) -> torch.nn.Module:
+    model = fx.symbolic_trace(model)
+    new_graph = fx.Graph()
+    env = {}
+    for node in model.graph.nodes:
+        if node.op == 'call_function' and node.target in decomposition_rules:
+            tracer = fx.proxy.GraphAppendingTracer(new_graph)
+            proxy_args = [fx.Proxy(env[x.name], tracer=tracer) if isinstance(x, fx.Node) else x for x in node.args]
+            new_node = decomposition_rules[node.target](*proxy_args).node
```

Comment:
For anything that's not an operator (including `addmm` here`, we might have a callsite with kwargs as well. I think this should section should look like:

```
proxy_args = map_arg(node.args, lambda n: Proxy(env[n.name]))
proxy_kwargs = map_arg(node.kwargs, lambda n: Proxy(env[n.name]))
new_node = decomposition_rules[node.target](*proxy_args, **proxy_kwargs).node
```